sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/SchemaAndRowProto.java

```diff
@@ -0,0 +1,166 @@
+  private final StructType tableSchema;
+  private final Row row;
+  TableRow tableRow;
+
+  public SchemaAndRowProto(StructType tableSchema, Row row) {
+    this.tableSchema = checkNotNull(tableSchema, "tableSchema");
+    this.row = checkNotNull(row, "row");
+  }
+
+  public Row getRow() {
+    return row;
+  }
+
+  public StructType getSchema() {
+    return tableSchema;
+  }
+
+  /**
+   * Gets a field value based on field name.
+   * Note: The implementation of this function will convert the entire row to {@link TableRow}
+   * and use the field name to get the value out of TableRow, based on the assumption that the
+   * most efficient way of accessing the subset of fields is to pass down the Field Selections
+   * to the read API.
+   *
+   * @param fieldName
+   * @return
+   */
+  public Object get(String fieldName) {
+    TableRow currentRow = getTableRow();
+    String[] subFieldNames = fieldName.split(".");
+    for (int i = 0; i < subFieldNames.length; i++) {
+      if (i == subFieldNames.length - 1) {
+        return currentRow.get(subFieldNames[i]);
+      } else {
+        // If there are further nested levels, the current field should contain a TableRow.
+        currentRow = (TableRow) (currentRow.get(subFieldNames[i]));
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Converts a {@class Row} that returned from ParallelRead API to {@class TableRow}
+   * which is a map of {fieldname, value} to be consumed by Data Flow API.
+   */
+  public TableRow getTableRow() {
+    if (tableRow != null) {
+      return tableRow;
+    }
+    tableRow = new TableRow();
+    List<RowOuterClass.Value> rowFields = row.getValue().getFieldsList();
+
+    checkState(rowFields.size() == tableSchema.getFieldsCount(),
+        "Expected that the row has the same number of cells %s as"
+            + " fields in the schema %s",
+        rowFields.size(), tableSchema.getFieldsCount());
+
+    for (int i = 0; i < tableSchema.getFieldsCount(); i++) {
+      // Convert the object in this cell to the Java type corresponding to its type in the schema.
+      Object convertedValue =
+          getFieldValue(tableSchema.getFields(i).getFieldType(), rowFields.get(i));
+
+      String fieldName = tableSchema.getFields(i).getFieldName();
+
+      if (convertedValue == null) {
+        // BigQuery does not include null values when the export operation (to JSON) is used.
+        // To match that behavior, BigQueryTableRowiterator, and the DirectPipelineRunner,
+        // intentionally omits columns with null values.
+        continue;
+      }
+
+      tableRow.set(fieldName, convertedValue);
+    }
+    return tableRow;
+  }
+
+  /**
+   * Convert parallel read field value {@link RowOuterClass.Value} into a object.
+   */
+  @Nullable
+  private Object getFieldValue(RowOuterClass.Type fieldType, RowOuterClass.Value fieldValue) {
+    // TODO: This check always evaluates to true, and so it breaks parsing.
+    // What's the right way to handle null values?
```

Comment:
Please clarify this. Is this a known bug / case we can't handle? In that case it's better to fail explicitly, than produce wrong data.